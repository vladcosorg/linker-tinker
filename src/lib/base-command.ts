import { Command, Flags } from '@oclif/core'

import type { AllContextOptions } from '@/lib/context'
import { enableDebug } from '@/lib/debug'

import type { Interfaces } from '@oclif/core'
import type { ListrRendererValue } from 'listr2'

export type Flags<T extends typeof Command> = Interfaces.InferredFlags<
  (typeof BaseCommand)['baseFlags'] & T['flags']
>
// eslint-disable-next-line unicorn/prevent-abbreviations
export type Args<T extends typeof Command> = Interfaces.InferredArgs<T['args']>

export abstract class BaseCommand<T extends typeof Command> extends Command {
  // define flags that can be inherited by any command that extends BaseCommand
  static override baseFlags = {
    verbose: Flags.boolean({
      char: 'v',
      description: 'Display a verbose action list',
      default: false,
    }),
    debug: Flags.boolean({
      char: 'd',
      default: false,
      description:
        'Enables the verbose mode and displays additional debug info that may help you',
    }),
  }

  protected flags!: Flags<T>
  protected args!: Args<T>

  public override async init(): Promise<void> {
    await super.init()

    console.log(super.ctor.baseFlags)
    const { args, flags } = await this.parse({
      flags: {
        ...this.ctor.flags,
        ...super.ctor.baseFlags,
      },
      args: this.ctor.args,
      strict: this.ctor.strict,
    })
    this.flags = flags as Flags<T>
    this.args = args as Args<T>

    if (this.flags.debug) {
      enableDebug()
    }
  }

  protected getRendererType(): ListrRendererValue {
    return this.flags.verbose || this.flags.debug ? 'simple' : 'default'
  }

  protected createContext<O extends Partial<AllContextOptions>>(context: O) {
    return {
      debug: this.flags.debug,
      isExiting: false,
      ...context,
    }
  }
}
