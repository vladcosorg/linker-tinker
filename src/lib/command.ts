import process from 'node:process'

import { Args } from '@oclif/core'

export function getInputArgs() {
  return {
    from: Args.string({
      description: 'Path to a package that you want to link',
      required: true,
    }),
    to: Args.string({
      description: `Target package. If not specified, it will use the current cwd`,
      required: false,
      default: process.cwd(),
    }),
  }
}
