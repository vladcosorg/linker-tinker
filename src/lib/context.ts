import type { ListrTask } from 'listr2'

export interface Context {
  sourcePackagePath: string
  targetPackagePath: string
  intermediateCacheDirectory: string
  syncPaths: string[] | string
  runWatcherScript: string | undefined
  debug: boolean
  bidirectionalSync: boolean
  isExiting: boolean
  skipWatch: boolean
  noSymlink: boolean
  watchAll: boolean
  pendingBidirectionalUpdates: { fromSource: string[]; toSource: string[] }
  dependentPackageName: string
  onlyAttach: boolean
  foregroundWatcher: boolean
  rollbackQueue: ListrTask[]
}

export type RequiredContext<T extends keyof Context> = Required<
  Pick<Context, T>
>

export type BaseContext = RequiredContext<'debug' | 'isExiting'>

export function createSubcontext<C extends Context, T extends Array<keyof C>>(
  contextKeys: T,
  context: C,
): Required<Pick<C, T[number]>> {
  return new Proxy(context, {
    get(target, name: string, receiver) {
      if (!contextKeys.includes(name) || !Reflect.has(target, name)) {
        throw new Error(`Getting non-existent property '${name}'`)
      }

      return Reflect.get(target, name, receiver)
    },
    set(target, name, value, receiver) {
      if (!Reflect.has(target, name)) {
        throw new Error(
          `Setting non-existent property '${name}', initial value: ${value}`,
        )
      }

      return Reflect.set(target, name, value, receiver)
    },
  }) as unknown as Required<Pick<C, T[number]>>
}
