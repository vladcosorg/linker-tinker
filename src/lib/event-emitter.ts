import { TypedEmitter } from 'tiny-typed-emitter'

interface Events {
  exit: () => Promise<void> | void
  exitImmediately: () => Promise<void> | void
}
export const eventBus = new TypedEmitter<Events>()
