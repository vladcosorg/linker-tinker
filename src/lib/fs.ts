import path from 'node:path'

import getCacheDir from 'cachedir'
import jetpack from 'fs-jetpack'

export async function removeFileAndContainingDirectoryIfEmpty(
  filePath: string,
): Promise<void> {
  const fileDirectory = path.dirname(filePath)
  const folderFiles = await jetpack.listAsync(path.dirname(filePath))
  if (folderFiles === undefined) {
    return
  }

  if (folderFiles.length === 1 && folderFiles[0] === path.basename(filePath)) {
    await jetpack.removeAsync(filePath)
    folderFiles.length = 0
  }

  if (folderFiles.length === 0) {
    await jetpack.removeAsync(fileDirectory)
  }
}

async function createCacheDirectory(cacheDirectory: string): Promise<void> {
  if (await jetpack.existsAsync(cacheDirectory)) {
    return
  }

  await jetpack.dirAsync(cacheDirectory)
}

const createdDirectories: Record<string, string> = {}
export async function getGlobalCacheDirectory(name: string): Promise<string> {
  let cachedCacheDirectory = createdDirectories[name]

  if (!cachedCacheDirectory) {
    cachedCacheDirectory = getCacheDir(name)
    await createCacheDirectory(cachedCacheDirectory)
  }

  return cachedCacheDirectory
}
