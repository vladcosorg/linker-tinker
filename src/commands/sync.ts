import path from 'node:path'
import process from 'node:process'

import { Args, Command, Flags } from '@oclif/core'
import chalk from 'chalk'

import { enableDebug } from '@/lib/debug'
import { runTasks } from '@/lib/sync/tasks'

export default class Sync extends Command {
  static override description = 'Link and sync a package as a dependency'

  static override examples = [
    {
      description: 'Force the command to execute',
      command: '<%= config.bin %> <%= command.id %> --force',
    },
  ]

  static override args = {
    from: Args.string({
      description: 'Path to a package that you want to link',
      required: true,
    }),
    to: Args.string({
      description: `Target package. If not specified, it will use the current cwd`,
      required: false,
      default: process.cwd(),
    }),
  }

  static override flags = {
    verbose: Flags.boolean({
      char: 'v',
      description: 'Display a verbose action list',
      default: false,
    }),
    noSymlink: Flags.boolean({
      char: 'n',
      description: 'Do not use symlink',
      default: false,
    }),
    debug: Flags.boolean({
      char: 'd',
      default: false,
      description:
        'Enables the verbose mode and displays additional debug info that may help you',
    }),
    bidirectionalSync: Flags.boolean({
      char: 'b',
      default: false,
      description: 'Enabled bidirectional sync',
    }),
    skipWatch: Flags.boolean({
      char: 's',
      default: false,
      hidden: true,
    }),
    watchAll: Flags.boolean({
      char: 'a',
      default: false,
      description: `Sync  all the changes in the ${chalk.bold(
        'dependent',
      )} package. By default, this utility will watch only for the paths returned by ${chalk.green(
        '`npm pack`',
      )}`,
    }),
    watcherScript: Flags.string({
      char: 'w',
      description:
        'Name of the watcher script in the dependent package.json to be executed. It will fail if the script prematurely or does not exist.',
    }),
  }

  async run(): Promise<void> {
    const { args, flags } = await this.parse(Sync)

    if (flags.debug) {
      enableDebug()
    }

    await runTasks({
      renderer: flags.verbose || flags.debug ? 'simple' : 'default',
      ctx: {
        skipWatch: flags.skipWatch,
        noSymlink: flags.noSymlink,
        isExiting: false,
        sourcePackagePath: path.resolve(args.from),
        targetPackagePath: path.resolve(args.to),
        syncPaths: path.resolve(args.from),
        runWatcherScript: flags.watcherScript,
        debug: flags.debug,
        bidirectionalSync: flags.bidirectionalSync,
        watchAll: flags.watchAll,
        pendingBidirectionalUpdates: { fromSource: [], toSource: [] },
        dependentPackageName: '',
        intermediatePackagePath: '',
      },
    })
  }
}
