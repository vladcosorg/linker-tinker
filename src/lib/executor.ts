import { execa } from 'execa'
import { createPromiseMixin } from 'promise-supplement'

import { debugConsole } from '@/lib/debug'
import { eventBus } from '@/lib/event-emitter'
import type { dependencyTypes as dependencyTypeList } from '@/lib/misc'

import type { ExecaChildProcess } from 'execa'

export const cancellableExeca = function (
  ...parameters: Parameters<typeof execa>
) {
  debugConsole.log(parameters)

  const originalChild = execa(...(parameters as Parameters<typeof execa>))
  const child = createPromiseMixin(originalChild.then(), originalChild)

  const cancelHandler = () => {
    child.cancel()
  }

  eventBus.once('exitImmediately', cancelHandler)

  return (
    child
      .then((result) => {
        debugConsole.log(result.command)
        return result
      })
      // .catch((error) => {
      //   debugConsole.log(error)
      //   throw error
      // })
      .finally(() => {
        eventBus.removeListener('exitImmediately', cancelHandler)
      })
  )

  // void (async () => {
  //   try {
  //     const result = await child
  //     debugConsole.log(result.command)
  //   } catch (error) {
  //     debugConsole.log(error)
  //   } finally {
  //     eventBus.removeListener('exitImmediately', cancelHandler)
  //   }
  // })()
} as unknown as typeof execa

const activeProcesses: Record<string, ExecaChildProcess> = {}

export function abortableExeca(
  id: string,
  command: string,
  commandArguments?: readonly string[],
  options?: execa.Options,
) {
  if (id in activeProcesses) {
    activeProcesses[id]?.cancel()
    delete activeProcesses[id]
  }

  // console.log(
  //   'aaa1',
  //   util.inspect(promise, { depth: 1, colors: true }),
  //   util.inspect(promise.then(), { depth: 1, colors: true }),
  // )

  const promise = cancellableExeca(command, commandArguments, {
    ...options,
  })

  promise.catch((error) => {
    if (!error.isCanceled) {
      throw error
    }

    return error
  })

  activeProcesses[id] = promise
  return promise
}

export type PackageConfig = {
  dependencyType?: (typeof dependencyTypeList)[number]
  versionRange?: string
} | null

export function runNpmInstall(
  rootPackagePath: string,
  dependencyName: string,
  config: PackageConfig = {},
): ExecaChildProcess {
  const dependencyTypeFlags: Record<
    (typeof dependencyTypeList)[number],
    string
  > = {
    dependencies: '--save-prod',
    devDependencies: '--save-dev',
  }

  const options = [
    'install',
    config?.versionRange
      ? `${dependencyName}@${config.versionRange}`
      : dependencyName,
    '--install-links',
    '--no-audit',
    '--no-fund',
    '--ignore-scripts',
  ]

  if (config?.dependencyType) {
    options.push(dependencyTypeFlags[config.dependencyType])
  }

  return cancellableExeca('npm', options, {
    cwd: rootPackagePath,
    all: true,
    shell: true,
  })
}

export function runNpmLink(
  rootPackagePath: string,
  dependencyName: string,
  {
    versionRange,
    dependencyType,
  }: {
    dependencyType?: (typeof dependencyTypeList)[number]
    versionRange?: string
  } = {},
): ExecaChildProcess {
  const dependencyTypeFlags: Record<
    (typeof dependencyTypeList)[number],
    string
  > = {
    dependencies: '--save-prod',
    devDependencies: '--save-dev',
  }
  const options = [
    'link',
    versionRange ? `${dependencyName}@${versionRange}` : dependencyName,
    '--install-links',
  ]

  if (dependencyType) {
    options.push(dependencyTypeFlags[dependencyType])
  }

  return cancellableExeca('npm', options, {
    cwd: rootPackagePath,
    all: true,
    shell: true,
  })
}

export function runNpmUninstall(
  rootPackagePath: string,
  dependencyName: string,
  save = true,
): ExecaChildProcess {
  return cancellableExeca(
    'npm',
    ['uninstall', dependencyName, save ? '--save' : '--no-save'],
    {
      cwd: rootPackagePath,
      all: true,
    },
  )
}

export function runNpmInstallRoot(rootPackagePath: string): ExecaChildProcess {
  return cancellableExeca(
    'npm',
    ['install', '--no-audit', '--no-fund', '--ignore-scripts'],
    {
      cwd: rootPackagePath,
      all: true,
    },
  )
}
