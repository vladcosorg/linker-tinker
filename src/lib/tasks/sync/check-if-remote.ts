import { homedir } from 'node:os'
import path from 'node:path'

import chalk from 'chalk'
import fs from 'fs-extra'

import { cloneRepo, runNpmInstallRoot } from '@/lib/executor'
import type { PickContext } from '@/lib/tasks'
import { createTask } from '@/lib/tasks'

type LocalContext = PickContext<'sourcePackagePath'>
export const checkIfRemote = createTask((context: LocalContext) => ({
  title: 'Checking if it is a remote package',
  task: async (_, task) => {
    const startsWithValidName = /^[\d@a-z]/.test(context.sourcePackagePath)

    if (!startsWithValidName) {
      task.title += `${chalk.green('[NO]')}`
      context.sourcePackagePath = path.resolve(context.sourcePackagePath)
      return
    }

    const response = await fetch(
      `https://registry.npmjs.org/${context.sourcePackagePath}/latest`,
    )
    const jsonResponse = (await response.json()) as
      | {
          name?: string
          repository?: { url?: string }
        }
      | undefined
    const url = jsonResponse?.repository?.url

    if (!url) {
      throw new Error(
        `The package ${context.sourcePackagePath} does not have a source repo specified`,
      )
    }

    const repo = url?.match(/github\.com\/(.*)\.git/)
    const fullRepo = `git@github.com:${repo[1]}.git`

    context.sourcePackagePath = path.join(
      homedir(),
      'linker-tinker',
      context.sourcePackagePath,
    )
    try {
      const stat = await fs.lstat(context.sourcePackagePath)

      if (!stat?.isDirectory()) {
        throw new Error('The provided path is not a directory')
      }

      task.title = `Using the cloned repository`
    } catch {
      task.title = `Cloning the repo ${chalk.bold(fullRepo)}`
      const process = cloneRepo(fullRepo, context.sourcePackagePath)
      process.all?.pipe(task.stdout())
      await process
    }

    const npmProcess = runNpmInstallRoot(context.sourcePackagePath)
    task.output = 'Installing the repo local repository'
    npmProcess.all?.pipe(task.stdout())
    await npmProcess
  },
}))
