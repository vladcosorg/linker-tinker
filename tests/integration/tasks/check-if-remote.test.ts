import { Listr } from 'listr2'
import { expect, test, vi } from 'vitest'

import { cloneRepo, runNpmInstallRoot } from '../../../src/lib/executor'
import { checkIfRemote } from '../../../src/lib/tasks/sync/check-if-remote'
import { activateProcessMock, recursiveJson } from '../../snaphot-util'
import { getFsHelpers } from '../../unit/helpers'

import type { ProcessMockTestContext } from '../../snaphot-util'

const { temporaryCwd } = getFsHelpers()
activateProcessMock()
vi.mock('../../../src/lib/executor', async () => {
  const actual = await vi.importActual('../../../src/lib/executor')
  return {
    ...actual,
    cloneRepo: vi.fn(() => vi.fn()),
    runNpmInstallRoot: vi.fn(() => vi.fn()),
  }
})

async function runTask(...parameters: Parameters<typeof checkIfRemote>) {
  return new Listr([checkIfRemote(...parameters)], { renderer: 'test' }).run()
}

vi.mock('node:os', () => ({
  homedir() {
    return temporaryCwd.path()
  },
}))

test<ProcessMockTestContext>('should clone the remote repo and install it', async ({
  stderr,
  stdout,
}) => {
  await runTask({ sourcePackagePath: 'chalk' })
  expect(vi.mocked(cloneRepo)).toHaveBeenCalledWith(
    'git@github.com:chalk/chalk.git',
    temporaryCwd.path('linker-tinker/chalk'),
  )
  expect(vi.mocked(runNpmInstallRoot)).toHaveBeenCalledWith(
    temporaryCwd.path('linker-tinker/chalk'),
  )

  expect(recursiveJson(stdout.mock.calls)).toMatchSnapshot()
  expect(recursiveJson(stderr.mock.calls)).toMatchSnapshot()
})

test<ProcessMockTestContext>('should reuse an existing repo if it was already cloned and then install it', async ({
  stderr,
  stdout,
}) => {
  await runTask({ sourcePackagePath: 'chalk' })
  temporaryCwd.dir('linker-tinker/chalk')
  await runTask({ sourcePackagePath: 'chalk' })

  expect(vi.mocked(cloneRepo)).toHaveBeenCalledOnce()
  expect(vi.mocked(runNpmInstallRoot)).toHaveBeenCalledTimes(2)

  expect(recursiveJson(stdout.mock.calls)).toMatchSnapshot()
  expect(recursiveJson(stderr.mock.calls)).toMatchSnapshot()
})

test<ProcessMockTestContext>('should skip if the provided package name is a path', async ({
  stderr,
  stdout,
}) => {
  await runTask({ sourcePackagePath: '/path/to/repo' })
  expect(vi.mocked(cloneRepo)).toHaveBeenCalledTimes(0)
  expect(vi.mocked(runNpmInstallRoot)).toHaveBeenCalledTimes(0)

  expect(recursiveJson(stdout.mock.calls)).toMatchSnapshot()
  expect(recursiveJson(stderr.mock.calls)).toMatchSnapshot()
})

test('should throw error if an invalid package is specified', async () => {
  await expect(async () =>
    runTask({ sourcePackagePath: 'foo-bar-vladcos' }),
  ).rejects.toThrowErrorMatchingSnapshot()
})

test('should throw error if a package without information about repository is specified', async () => {
  await expect(async () =>
    runTask({ sourcePackagePath: '@vladcos/prettier-config' }),
  ).rejects.toThrowErrorMatchingSnapshot()
})
