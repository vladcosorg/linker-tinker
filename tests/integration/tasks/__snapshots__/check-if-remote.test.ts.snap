// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should clone the remote repo and install it 1`] = `
[
  [
    {
      "data": "STARTED",
      "event": "STATE",
      "task": {
        "hasFailed": false,
        "hasFinalized": false,
        "hasReset": false,
        "hasRolledBack": false,
        "hasSubtasks": false,
        "hasTitle": true,
        "isCompleted": false,
        "isEnabled": true,
        "isPaused": false,
        "isPending": true,
        "isPrompt": false,
        "isRetrying": false,
        "isRollingBack": false,
        "isSkipped": false,
        "isStarted": true,
        "path": [
          "Checking if it is a remote package",
        ],
        "title": "Checking if it is a remote package",
      },
    },
  ],
  [
    {
      "data": "Cloning the repo [1mgit@github.com:chalk/chalk.git[22m",
      "event": "TITLE",
      "task": {
        "hasFailed": false,
        "hasFinalized": false,
        "hasReset": false,
        "hasRolledBack": false,
        "hasSubtasks": false,
        "hasTitle": true,
        "isCompleted": false,
        "isEnabled": true,
        "isPaused": false,
        "isPending": true,
        "isPrompt": false,
        "isRetrying": false,
        "isRollingBack": false,
        "isSkipped": false,
        "isStarted": true,
        "path": [
          "Checking if it is a remote package",
        ],
        "title": "Cloning the repo [1mgit@github.com:chalk/chalk.git[22m",
      },
    },
  ],
  [
    {
      "data": "Installing the repo local repository",
      "event": "OUTPUT",
      "task": {
        "hasFailed": false,
        "hasFinalized": false,
        "hasReset": false,
        "hasRolledBack": false,
        "hasSubtasks": false,
        "hasTitle": true,
        "isCompleted": false,
        "isEnabled": true,
        "isPaused": false,
        "isPending": true,
        "isPrompt": false,
        "isRetrying": false,
        "isRollingBack": false,
        "isSkipped": false,
        "isStarted": true,
        "path": [
          "Checking if it is a remote package",
        ],
        "title": "Cloning the repo [1mgit@github.com:chalk/chalk.git[22m",
      },
    },
  ],
  [
    {
      "data": "COMPLETED",
      "event": "STATE",
      "task": {
        "hasFailed": false,
        "hasFinalized": true,
        "hasReset": false,
        "hasRolledBack": false,
        "hasSubtasks": false,
        "hasTitle": true,
        "isCompleted": true,
        "isEnabled": true,
        "isPaused": false,
        "isPending": false,
        "isPrompt": false,
        "isRetrying": false,
        "isRollingBack": false,
        "isSkipped": false,
        "isStarted": false,
        "path": [
          "Checking if it is a remote package",
        ],
        "title": "Cloning the repo [1mgit@github.com:chalk/chalk.git[22m",
      },
    },
  ],
]
`;

exports[`should clone the remote repo and install it 2`] = `[]`;

exports[`should reuse an existing repo if it was already cloned and then install it 1`] = `
[
  [
    {
      "data": "STARTED",
      "event": "STATE",
      "task": {
        "hasFailed": false,
        "hasFinalized": false,
        "hasReset": false,
        "hasRolledBack": false,
        "hasSubtasks": false,
        "hasTitle": true,
        "isCompleted": false,
        "isEnabled": true,
        "isPaused": false,
        "isPending": true,
        "isPrompt": false,
        "isRetrying": false,
        "isRollingBack": false,
        "isSkipped": false,
        "isStarted": true,
        "path": [
          "Checking if it is a remote package",
        ],
        "title": "Checking if it is a remote package",
      },
    },
  ],
  [
    {
      "data": "Cloning the repo [1mgit@github.com:chalk/chalk.git[22m",
      "event": "TITLE",
      "task": {
        "hasFailed": false,
        "hasFinalized": false,
        "hasReset": false,
        "hasRolledBack": false,
        "hasSubtasks": false,
        "hasTitle": true,
        "isCompleted": false,
        "isEnabled": true,
        "isPaused": false,
        "isPending": true,
        "isPrompt": false,
        "isRetrying": false,
        "isRollingBack": false,
        "isSkipped": false,
        "isStarted": true,
        "path": [
          "Checking if it is a remote package",
        ],
        "title": "Cloning the repo [1mgit@github.com:chalk/chalk.git[22m",
      },
    },
  ],
  [
    {
      "data": "Installing the repo local repository",
      "event": "OUTPUT",
      "task": {
        "hasFailed": false,
        "hasFinalized": false,
        "hasReset": false,
        "hasRolledBack": false,
        "hasSubtasks": false,
        "hasTitle": true,
        "isCompleted": false,
        "isEnabled": true,
        "isPaused": false,
        "isPending": true,
        "isPrompt": false,
        "isRetrying": false,
        "isRollingBack": false,
        "isSkipped": false,
        "isStarted": true,
        "path": [
          "Checking if it is a remote package",
        ],
        "title": "Cloning the repo [1mgit@github.com:chalk/chalk.git[22m",
      },
    },
  ],
  [
    {
      "data": "COMPLETED",
      "event": "STATE",
      "task": {
        "hasFailed": false,
        "hasFinalized": true,
        "hasReset": false,
        "hasRolledBack": false,
        "hasSubtasks": false,
        "hasTitle": true,
        "isCompleted": true,
        "isEnabled": true,
        "isPaused": false,
        "isPending": false,
        "isPrompt": false,
        "isRetrying": false,
        "isRollingBack": false,
        "isSkipped": false,
        "isStarted": false,
        "path": [
          "Checking if it is a remote package",
        ],
        "title": "Cloning the repo [1mgit@github.com:chalk/chalk.git[22m",
      },
    },
  ],
  [
    {
      "data": "STARTED",
      "event": "STATE",
      "task": {
        "hasFailed": false,
        "hasFinalized": false,
        "hasReset": false,
        "hasRolledBack": false,
        "hasSubtasks": false,
        "hasTitle": true,
        "isCompleted": false,
        "isEnabled": true,
        "isPaused": false,
        "isPending": true,
        "isPrompt": false,
        "isRetrying": false,
        "isRollingBack": false,
        "isSkipped": false,
        "isStarted": true,
        "path": [
          "Checking if it is a remote package",
        ],
        "title": "Checking if it is a remote package",
      },
    },
  ],
  [
    {
      "data": "Using the cloned repository",
      "event": "TITLE",
      "task": {
        "hasFailed": false,
        "hasFinalized": false,
        "hasReset": false,
        "hasRolledBack": false,
        "hasSubtasks": false,
        "hasTitle": true,
        "isCompleted": false,
        "isEnabled": true,
        "isPaused": false,
        "isPending": true,
        "isPrompt": false,
        "isRetrying": false,
        "isRollingBack": false,
        "isSkipped": false,
        "isStarted": true,
        "path": [
          "Checking if it is a remote package",
        ],
        "title": "Using the cloned repository",
      },
    },
  ],
  [
    {
      "data": "Installing the repo local repository",
      "event": "OUTPUT",
      "task": {
        "hasFailed": false,
        "hasFinalized": false,
        "hasReset": false,
        "hasRolledBack": false,
        "hasSubtasks": false,
        "hasTitle": true,
        "isCompleted": false,
        "isEnabled": true,
        "isPaused": false,
        "isPending": true,
        "isPrompt": false,
        "isRetrying": false,
        "isRollingBack": false,
        "isSkipped": false,
        "isStarted": true,
        "path": [
          "Checking if it is a remote package",
        ],
        "title": "Using the cloned repository",
      },
    },
  ],
  [
    {
      "data": "COMPLETED",
      "event": "STATE",
      "task": {
        "hasFailed": false,
        "hasFinalized": true,
        "hasReset": false,
        "hasRolledBack": false,
        "hasSubtasks": false,
        "hasTitle": true,
        "isCompleted": true,
        "isEnabled": true,
        "isPaused": false,
        "isPending": false,
        "isPrompt": false,
        "isRetrying": false,
        "isRollingBack": false,
        "isSkipped": false,
        "isStarted": false,
        "path": [
          "Checking if it is a remote package",
        ],
        "title": "Using the cloned repository",
      },
    },
  ],
]
`;

exports[`should reuse an existing repo if it was already cloned and then install it 2`] = `[]`;

exports[`should skip if the provided package name is a path 1`] = `
[
  [
    {
      "data": "STARTED",
      "event": "STATE",
      "task": {
        "hasFailed": false,
        "hasFinalized": false,
        "hasReset": false,
        "hasRolledBack": false,
        "hasSubtasks": false,
        "hasTitle": true,
        "isCompleted": false,
        "isEnabled": true,
        "isPaused": false,
        "isPending": true,
        "isPrompt": false,
        "isRetrying": false,
        "isRollingBack": false,
        "isSkipped": false,
        "isStarted": true,
        "path": [
          "Checking if it is a remote package",
        ],
        "title": "Checking if it is a remote package",
      },
    },
  ],
  [
    {
      "data": "Checking if it is a remote package[32m[NO][39m",
      "event": "TITLE",
      "task": {
        "hasFailed": false,
        "hasFinalized": false,
        "hasReset": false,
        "hasRolledBack": false,
        "hasSubtasks": false,
        "hasTitle": true,
        "isCompleted": false,
        "isEnabled": true,
        "isPaused": false,
        "isPending": true,
        "isPrompt": false,
        "isRetrying": false,
        "isRollingBack": false,
        "isSkipped": false,
        "isStarted": true,
        "path": [
          "Checking if it is a remote package",
        ],
        "title": "Checking if it is a remote package[32m[NO][39m",
      },
    },
  ],
  [
    {
      "data": "COMPLETED",
      "event": "STATE",
      "task": {
        "hasFailed": false,
        "hasFinalized": true,
        "hasReset": false,
        "hasRolledBack": false,
        "hasSubtasks": false,
        "hasTitle": true,
        "isCompleted": true,
        "isEnabled": true,
        "isPaused": false,
        "isPending": false,
        "isPrompt": false,
        "isRetrying": false,
        "isRollingBack": false,
        "isSkipped": false,
        "isStarted": false,
        "path": [
          "Checking if it is a remote package",
        ],
        "title": "Checking if it is a remote package[32m[NO][39m",
      },
    },
  ],
]
`;

exports[`should skip if the provided package name is a path 2`] = `[]`;

exports[`should throw error if a package without information about repository is specified 1`] = `"The package @vladcos/prettier-config does not have a source repo specified"`;

exports[`should throw error if an invalid package is specified 1`] = `"The package foo-bar-vladcos does not have a source repo specified"`;
